File: scripts/notion_to_content/indexes.py
from __future__ import annotations

import csv
import re
from pathlib import Path

from . import textutil

_ID_VIEW = re.compile(
    r"^(.+?)\s+([a-f0-9]{32})((?:_all(?:_[0-9]+)?)?)\.csv$",
    re.IGNORECASE,
)


def parse_csv_basename(
    name: str,
) -> tuple[str, str, str, str] | None:
    """
    Return (display_name, database_id_hex, view_key, file_stem) or None.
    view_key: 'default' | 'all' | 'all-1' | ...
    """
    m = _ID_VIEW.match(name)
    if not m:
        return None
    display_name, dbid, suff = m.group(1), m.group(2).lower(), m.group(3) or ""
    if not suff:
        view = "default"
    elif suff == "_all":
        view = "all"
    else:
        m2 = re.match(r"^_all_(\d+)$", suff)
        view = f"all-{m2.group(1)}" if m2 else suff.lstrip("_") or "default"
    return display_name, dbid, view, name


def index_out_filename(database_id: str, view: str) -> str:
    vid = database_id.lower()
    v = view.lower().replace("_", "-")
    return f"index-{vid}-{v}.md"


def read_csv_rows(path: Path) -> tuple[list[str], list[list[str]]]:
    with path.open(newline="", encoding="utf-8", errors="replace") as f:
        reader = csv.reader(f)
        rows = list(reader)
    if not rows:
        return [], []
    header = [textutil.strip_emojis(h) for h in rows[0]]
    header = textutil.disambiguate_headers(header)
    return header, rows[1:]

File: packages/notion-importer/src/indexes.ts
import { readFileSync } from "node:fs";
import { disambiguateHeaders, stripEmojis } from "./textutil";

const ID_VIEW =
  /^(.+?)\s+([a-f0-9]{32})((?:_all(?:_[0-9]+)?)?)\.csv$/i;

export interface ParsedCsvBasename {
  displayName: string;
  databaseId: string;
  viewKey: string;
  fileStem: string;
}

export function parseCsvBasename(name: string): ParsedCsvBasename | null {
  const m = ID_VIEW.exec(name);
  if (!m) return null;
  const displayName = m[1]!;
  const dbid = m[2]!.toLowerCase();
  const suff = m[3] ?? "";
  let view: string;
  if (!suff) view = "default";
  else if (suff === "_all") view = "all";
  else {
    const m2 = /^_all_(\d+)$/.exec(suff);
    view = m2 ? `all-${m2[1]}` : suff.replace(/^_/, "") || "default";
  }
  return { displayName, databaseId: dbid, viewKey: view, fileStem: name };
}

export function indexOutFilename(databaseId: string, view: string): string {
  const vid = databaseId.toLowerCase();
  const v = view.toLowerCase().replace(/_/g, "-");
  return `index-${vid}-${v}.md`;
}

/** Minimal CSV reader matching Python csv.reader on Notion exports. */
function parseCsv(text: string): string[][] {
  const rows: string[][] = [];
  let row: string[] = [];
  let field = "";
  let i = 0;
  let inQuotes = false;

  while (i < text.length) {
    const c = text[i]!;
    if (inQuotes) {
      if (c === '"') {
        if (text[i + 1] === '"') {
          field += '"';
          i += 2;
          continue;
        }
        inQuotes = false;
        i += 1;
        continue;
      }
      field += c;
      i += 1;
      continue;
    }
    if (c === '"') {
      inQuotes = true;
      i += 1;
      continue;
    }
    if (c === ",") {
      row.push(field);
      field = "";
      i += 1;
      continue;
    }
    if (c === "\r") {
      i += 1;
      continue;
    }
    if (c === "\n") {
      row.push(field);
      rows.push(row);
      row = [];
      field = "";
      i += 1;
      continue;
    }
    field += c;
    i += 1;
  }
  if (field.length > 0 || row.length > 0) {
    row.push(field);
    rows.push(row);
  }
  return rows;
}

export function readCsvRows(path: string): [string[], string[][]] {
  const text = readFileSync(path, { encoding: "utf-8" });
  const rows = parseCsv(text);
  if (rows.length === 0) return [[], []];
  let header = rows[0]!.map((h) => stripEmojis(h));
  header = disambiguateHeaders(header);
  return [header, rows.slice(1)];
}
